openedx/core/lib/block_structure/transformers.py

```diff
@@ -105,3 +100,48 @@ def is_collected_outdated(cls, block_structure):
             )
 
         return bool(outdated_transformers)
+
+    def transform(self, block_structure):
+        """
+        The given block structure is transformed by each transformer in the
+        collection, in the order that the transformers were added.
+        """
+        self._transform_with_filter(block_structure)
+        self._full_transform(block_structure)
+
+        # Prune the block structure to remove any unreachable blocks.
+        block_structure._prune_unreachable()  # pylint: disable=protected-access
+
+    def _transform_with_filter(self, block_structure):
+        """
+        Transforms the given block_structure using the transform_block_filter
+        method from the given transformers.
+        """
+        if not self._transformers['supports_filter']:
+            return
+
+        filters = []
+        for transformer in self._transformers['supports_filter']:
+            filters.extend(transformer.transform_block_filter(self.usage_info, block_structure))
+
+        combined_filters = functools.reduce(
+            self._filter_chain,
+            filters,
+            block_structure.create_universal_filter()
+        )
+        block_structure.filter_topological_traversal(combined_filters)
+
+    def _filter_chain(self, accumulated, additional):
+        """
+        Given two functions that take a block_key and return a boolean, yield
+        a function that takes a block key, and 'ands' the functions together
+        """
+        return lambda block_key: accumulated(block_key) and additional(block_key)
+
+    def _full_transform(self, block_structure):
```

Comment:
Nit: why choose to introduce a new term, "full_transform", that the reader needs to understand?  Continue with the term, "no_filter"?